Fire events for Lutron RadioRA2 keypad buttons

koreth opened the pull request:
## Description:

Add a binary sensor component for Lutron RadioRA2 keypad buttons so button presses can trigger actions in HA.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8164

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`. *(New file is already covered by an existing glob pattern)*

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

koreth committed:
Add binary sensor for Lutron RadioRA2 keypad buttons

Allow automations to be triggered from RadioRA2 keypads by exposing
each button as a binary sensor.

balloob commented on homeassistant/components/binary_sensor/lutron.py:
```diff
@@ -0,0 +1,84 @@
+from homeassistant.components.lutron import LUTRON_DEVICES
+
+DEPENDENCIES = ['lutron']
+REQUIREMENTS = ['pylutron==0.2.0']
```
No need for requirement, you're already depending on `lutron` as dependency

balloob commented on homeassistant/components/binary_sensor/lutron.py:
```diff
@@ -0,0 +1,84 @@
+"""
```
Please move this file to `homeassistant/components/lutron/binary_sensor.py`

balloob commented on homeassistant/components/binary_sensor/lutron.py:
```diff
@@ -0,0 +1,84 @@
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Lutron RadioRA2 binary sensor platform."""
+    if LUTRON_DEVICES not in hass.data:
```
This is not possible, because it means the component was not set up correctly, which would cause the platform to be never loaded.

balloob commented on homeassistant/components/binary_sensor/lutron.py:
```diff
@@ -0,0 +1,84 @@
+    """Set up the Lutron RadioRA2 binary sensor platform."""
+    if LUTRON_DEVICES not in hass.data:
+        return
+    if 'binary_sensor' not in hass.data[LUTRON_DEVICES]:
```
same

balloob commented on homeassistant/components/binary_sensor/lutron.py:
```diff
@@ -0,0 +1,84 @@
+
+    for (name, button) in hass.data[LUTRON_DEVICES]['binary_sensor']:
+        sensor = LutronButtonSensor(name, button, hass)
+        add_devices([sensor])
```
Don't call add_devices many times but instead call it once with a list of sensors.

balloob reviewed (changes requested):
This is the wrong approach. Buttons should not be represented as binary sensors. Instead, they should fire events.

koreth commented:
I agree that binary sensors aren't a great semantic fit (especially for the single-action buttons). What component type would be more suitable, though? Or are you suggesting that the buttons shouldn't be represented as components at all? Are there any examples of other integrations I should model this after?

Thanks for all the feedback! Happy to do what it takes to do this the right way.

MartinHjelmare commented:
Stateless devices should just fire custom events and not be part of an entity component. See the deconz component for a good example:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/deconz/gateway.py#L150-L176

koreth committed:
Remove binary sensor component; fire events directly instead.

koreth commented:
Reimplemented this to directly fire events rather than registering new components, so I'm marking the comments on the original implementation as resolved since that code no longer exists.

MartinHjelmare commented on homeassistant/components/lutron.py:
```diff
@@ -110,3 +113,45 @@ def name(self):
+
+        data = {CONF_ID: self._id, CONF_EVENT: action}
+
+        self._hass.bus.fire(self._event, data, EventOrigin.remote)
```
The origin is not remote, ie not from another home assistant instance. We can remove that argument.

MartinHjelmare commented on homeassistant/components/lutron.py:
```diff
@@ -105,3 +113,45 @@ def name(self):
+
+    def __init__(self, hass, keypad, button):
+        """Register callback for activity on the button."""
+        self._name = '{}: {}'.format(keypad.name, button.name)
```
`self._name` isn't used as an instance attribute. We can just store it as a local variable.

MartinHjelmare commented on homeassistant/components/lutron.py:
```diff
@@ -105,3 +113,45 @@ def name(self):
+        """Register callback for activity on the button."""
+        self._name = '{}: {}'.format(keypad.name, button.name)
+        self._hass = hass
+        self._button = button
```
This isn't used.

MartinHjelmare commented on homeassistant/components/lutron.py:
```diff
@@ -105,3 +113,45 @@ def name(self):
+        self._button = button
+        self._has_release_event = 'RaiseLower' in button.button_type
+        self._id = slugify(self._name)
+        self._event = 'lutron_{}'.format(CONF_EVENT)
```
This isn't a configuration key, so we shouldn't use `CONF_EVENT`. Just write `lutron_event` to the variable.

MartinHjelmare commented on homeassistant/components/lutron.py:
```diff
@@ -105,3 +113,45 @@ def name(self):
+
+        button.subscribe(self.button_callback, None)
+
+    @callback
```
We should remove this, since this callback won't be called in the event loop. It will be called by another thread.

MartinHjelmare commented on homeassistant/components/lutron.py:
```diff
@@ -105,3 +113,45 @@ def name(self):
+            # than for buttons where we expect a release event.
+            action = 'single'
+
+        data = {CONF_ID: self._id, CONF_EVENT: action}
```
Use `ATTR_` constants for event attribute keys.

koreth committed:
Address comments from code review

koreth commented:
Thanks for the thorough review! I've addressed all these comments. I also renamed the `event` attribute in the event data to `action` after noticing that calling it `event` makes it unusable with the filtering feature of AppDaemon's event listener API.

The deconz component, which I used as a model for my implementation, also has some of these problems. I'm happy to fix them there too but I would have no way to test my changes.

MartinHjelmare reviewed (approved):
Great!

sgrimm-sg commented:
@Kane610 To save you the trouble of looking through the individual bits of feedback, these are the things I fixed in the last revision of my PR that might need attention in deconz:

* I was using `CONF_EVENT` to construct the event type, which is inappropriate because it's not a configuration item; I changed the event type to a literal containing the word `event` instead.
* I passed `EventOrigin.remote` when I fired the event, which is inappropriate because the event doesn't originate from a remote HA instance. (I had assumed it meant the event originated outside of HA.)
* The callback function I passed to the subscription API in the underlying device was unnecessarily annotated with `@callback`; it wasn't called from the HA event loop.

Changing the `event` attribute to `action` just seemed like a good idea so people could pass, e.g., `action='single'` as a keyword argument to the event subscription API in AppDaemon apps to filter for just that action. You can't do `event='single'` because `event` is the argument for the event type. Obviously I didn't have backward compatibility to worry about, though.

koreth commented:
@Kane610 To save you the trouble of looking through the individual bits of feedback, these are the things I fixed in the last revision of my PR that might need attention in deconz:

* I was using `CONF_EVENT` to construct the event type, which is inappropriate because it's not a configuration item; I changed the event type to a literal containing the word `event` instead.
* I passed `EventOrigin.remote` when I fired the event, which is inappropriate because the event doesn't originate from a remote HA instance. (I had assumed it meant the event originated outside of HA.)
* The callback function I passed to the subscription API in the underlying device was unnecessarily annotated with `@callback`; it wasn't called from the HA event loop.

Changing the `event` attribute to `action` just seemed like a good idea so people could pass, e.g., `action='single'` as a keyword argument to the event subscription API in AppDaemon apps to filter for just that action. You can't do `event='single'` because `event` is the argument for the event type. Obviously I didn't have backward compatibility to worry about, though.

Kane610 commented:
Thanks for that great summary @koreth .
